feat(docs): added deployment page for GCP (#828)
kestra-io/docs: content/docs/09.administrator-guide/02.deployment/07.google-vm.md

1. Go to the database overview page and click on `Users` from the left-side navigation menu.
2. Click on `Add User Account`.
3. Put an appropriate username and password, and click on `Add`.

![db_users](/docs/administrator-guide/deployment/gcp-vm/db_users.png)

![db_user_creation](/docs/administrator-guide/deployment/gcp-vm/db_user_creation.png)

**Create Kestra database**

1. Go to the database overview page, and click on `Databases` from the left side navigation menu.
2. Click on `Create Database`.
3. Put an appropriate database name, and click on `Create`.

**Update Kestra configuration**

In the docker-compose configuration, edit the `datasources` property of the Kestra service in the following way:

```yaml
datasources:
  postgres:
  url: jdbc:postgresql://<your-db-external-endpoint>:5432/<db_name>
  driverClassName: org.postgresql.Driver
  username: <your-username>
  password: <your-password>
```

Because you now use the Cloud SQL service, you don't need the Postgres Docker service anymore. Remove it from the `docker-compose.yml` file.

In order for the changes to take effect, restart the docker services with `docker compose restart` or `docker compose up -d`.

> ## Step 5: Use AWS S3 for storage

Reviewer: ```suggestion
## Step 5: Use GCS for internal storage
```

---
title: Deployment on Google Cloud Platform (Compute Engine, Cloud SQL, Cloud Storage)
---

This guide provides instructions for deploying Kestra on Google Cloud Platform (GCP). We’ll use Compute Engine with Docker to host Kestra server, a PostgreSQL Cloud SQL database and Cloud Storage as storage backend.

Prerequisites:
* basic knowledge about using a command line interface
* basic knowledge about Compute Engine, Cloud Storage and PostgreSQL.

## Step 1: Create a VM instance & install Docker

> First, create an VM instance in Compute Engine. To do so, [go to the GCP console and choose Compute Engine](https://console.cloud.google.com/compute/instances).
> 1. Click on the `Create Instance` button at the top.
> 2. Give a name to your instance.
> 3. Choose an appropriate Region and Zone.
> 4. Choose `General Purpose` machine of `E2` series.
> 5. Machine type: Kestra needs at least 4GiB Memory and 2vCPU to run correctly. Choosing `Preset` machine type `e2-standard-2` is a good starting point.
> 6. Click on `Change` in the "Boot Disk" section, as we would like to change the image.
> 7. Under "Public Images" tab, choose `Ubuntu` as the operating system, and `Ubuntu 22.04 LTS` version.
> 8. Continue with "Allow default access" access scope, and select "Allow HTTPS traffic" in the Firewall section.
> 
> ![vm creation_1](/docs/administrator-guide/deployment/gcp-vm/vm_setup1.png)
> 
> ![vm creation 2](/docs/administrator-guide/deployment/gcp-vm/vm_setup2.png)
> 
> ![change_boot_disk_image](/docs/administrator-guide/deployment/gcp-vm/vm_setup3.png)
> 
> ![vm_creation_3](/docs/administrator-guide/deployment/gcp-vm/vm_setup4.png)
> 
> You can now click on "Create" and wait a few seconds for the VM instance to be up and running.
> 
> Then, click on the `SSH` button on the right of the VM instance details to SSH into the VM instance terminal. Click on the `Authorize` button in the pop-up to authorize the SSH connection into the VM instance.
> 
> ![ssh_into_vm](/docs/administrator-guide/deployment/gcp-vm/ssh_into_vm.png)
> 
> Kestra can be run directly from the .jar binary or thanks to Docker. We’ll use docker here for quicker setup:
> 1. Install Docker on the EC2 instance. [You can find the last updated instruction on Docker website](https://docs.docker.com/engine/install/ubuntu/).
> 2. [Install docker-compose](https://docs.docker.com/compose/install/).
> 
> To check your installation run `docker version` and `docker-compose version`. You're now ready to download and launch Kestra server.

Reviewer: ```suggestion
First, create a VM instance using the Compute Engine. To do so, [go to the GCP console and choose Compute Engine](https://console.cloud.google.com/compute/instances).
1. Click on the `Create Instance` button at the top.
2. Give a name to your instance.
3. Choose an appropriate Region and Zone.
4. Choose the `General Purpose` machine of the `E2` series.
5. Machine type: Kestra needs at least 4GiB Memory and 2vCPU to run correctly. Choosing the `Preset` machine type `e2-standard-2` is a good starting point.
6. Click on `Change` in the "Boot Disk" section, as we would like to change the image.
7. Under the "Public Images" tab, choose `Ubuntu` as the operating system and the `Ubuntu 22.04 LTS` version.
8. Continue with the "Allow default access" access scope, and select "Allow HTTPS traffic" in the Firewall section.

![vm creation_1](/docs/administrator-guide/deployment/gcp-vm/vm_setup1.png)

![vm creation 2](/docs/administrator-guide/deployment/gcp-vm/vm_setup2.png)

![change_boot_disk_image](/docs/administrator-guide/deployment/gcp-vm/vm_setup3.png)

![vm_creation_3](/docs/administrator-guide/deployment/gcp-vm/vm_setup4.png)

You can now click on "Create" and wait a few seconds for the VM instance to be up and running.

Then, click on the `SSH` button on the right of the VM instance details to SSH into the VM instance terminal. Click on the `Authorize` button in the pop-up to authorize the SSH connection into the VM instance.

![ssh_into_vm](/docs/administrator-guide/deployment/gcp-vm/ssh_into_vm.png)

Kestra can be run directly from the .jar binary or thanks to Docker. We’ll use docker here for quicker setup:
1. Install Docker on the EC2 instance. [You can find the last updated instruction on the Docker website](https://docs.docker.com/engine/install/ubuntu/).
2. [Install docker-compose](https://docs.docker.com/compose/install/).

To check your installation, run `docker version` and `docker-compose version`. You're now ready to download and launch the Kestra server.
```

## Step 2: Download and Run Kestra

> Download the official Docker-Compose file: `curl -o docker-compose.yml https://raw.githubusercontent.com/kestra-io/kestra/develop/docker-compose.yml`
> 
> Use editor such as Vim to modify the `docker-compose.yml` and set basic authentication to true, and configure your basic authentication credentials to secure your Kestra instance.
> 
> ```yaml
> kestra:
>   server:
>     basic-auth:
>       enabled: true
>       username: admin@kestra.io # it must be a valid email address
>       password: kestra
> ```
> 
> Then use the following command to start Kestra server: `sudo docker-compose up -d`
> 
> 
> ## Step 3: Allow external traffic
> 
> Kestra is now running and the Kestra server exposes traffic on the 8080 port. To connect through your web browser, we need to allow the traffic from port 8080 to the VM.
> 
> 1. Go to the Compute Engine console, and click on the three dots to the utmost right on the VM instance row. From the dropdown menu, click on `View Network Details`.
> 2. On the Network Interface Details page, click on `Firewall` from the left side navigation menu.
> 3. On the Firewall Policies page, click on `Create Firewall Rule` button at the top of the page.
> 4. Give an appropriate name like `kestra-ui-port`, and an appropriate description to the Firewall Rule.
> 5. In the targets, select `All instance in the network`, select `IPv4 ranges` as the Source filter, and provide appropriate IPv4 address or range in the Source IPv4 ranges.
>    If you want to only allow traffic coming from your IP address, put your own IP. If you want to make it open to the entire Internet, put the value `0.0.0.0/0`.
> 6. Under Protocols and ports section, in the Specified protocols and ports, select `TCP` and enter `8080` in the Ports.
> 7. Click on `Create` at the bottom of the page.

Reviewer: ```suggestion
Download the official Docker-Compose file: `curl -o docker-compose.yml https://raw.githubusercontent.com/kestra-io/kestra/develop/docker-compose.yml`

Use an editor such as Vim to modify the `docker-compose.yml`, set basic authentication to true, and configure your basic authentication credentials to secure your Kestra instance.

```yaml
kestra:
  server:
    basic-auth:
      enabled: true
      username: admin@kestra.io # it must be a valid email address
      password: kestra
```

Then, use the following command to start the Kestra server: `sudo docker-compose up -d`

## Step 3: Allow external traffic

Kestra is now running, and the Kestra server exposes traffic on the 8080 port. To connect through your web browser, we need to allow the traffic from port 8080 to the VM.

1. Go to the Compute Engine console and click on the three dots to the utmost right on the VM instance row. From the dropdown menu, click on `View Network Details`.
2. On the Network Interface Details page, click on `Firewall` from the left-side navigation menu.
3. On the Firewall Policies page, click on the `Create Firewall Rule` button at the top of the page.
4. Give an appropriate name, such as `kestra-ui-port`, and an appropriate description for the Firewall Rule.
5. In the targets, select `All instance in the network`, select `IPv4 ranges` as the Source filter, and provide the appropriate IPv4 address or range in the Source IPv4 ranges.
   If you want to only allow traffic coming from your IP address, put your own IP. If you want to make it open to the entire Internet, put the value `0.0.0.0/0`.
6. Under the Protocols and ports section, in the Specified protocols and ports, select `TCP` and enter `8080` in the Ports.
7. Click on `Create` at the bottom of the page.
```

![vm_network_details_option](/docs/administrator-guide/deployment/gcp-vm/vm_network_details_option.png)

![vm_network_interface_details](/docs/administrator-guide/deployment/gcp-vm/vm_network_interface_details.png)

![vm_firewall_policies](/docs/administrator-guide/deployment/gcp-vm/vm_firewall_policies.png)

![vm_create_firewall_rule](/docs/administrator-guide/deployment/gcp-vm/vm_create_firewall_rule.png)

::alert{type="warning"}

> Note that if you haven't set up basic authentication in the previous step, your Kestra instance will be publicly available to anyone without any access restriction.
> ::
> 
> You can now access your Kestra instance and create, edit and run Flows.
> 
> 
> ## Step 4: Use Cloud SQL PostgreSQL as a database backend
> 
> This first installation relies on a PostgreSQL database running alongside the Kestra server - on the VM instance (see the PostgreSQL service running thanks to the docker-compose).
> 
> For a simple PoC, you can keep that Postgres database running in Docker.
> 
> However, for a production-grade installation, we recommend a managed database service such as Cloud SQL.
> 
> **Create a Cloud SQL database**
> 
> 1. Go to the [Cloud SQL console](https://console.cloud.google.com/sql/instances).
> 2. Click on `Choose PostgreSQL` (Kestra also supports MySQL, but PostgreSQL is recommended).
> 3. Put an appropriate Instance ID, and password for the admin user `postgres`.
> 4. Select the latest PostgreSQL version from the dropdown.
> 5. Choose `Enterprise Plus` or `Enterprise` edition based on your requirement.
> 6. Choose an appropriate preset among `Production`, `Development` or `Sandbox` as per your requirement.
> 7. Choose appropriate region and zonal availability.
> 8. Hit create and wait for completion.

Reviewer: ```suggestion
Note that if you haven't set up basic authentication in the previous step, your Kestra instance will be publicly available to anyone without any access restriction.
::

You can now access your Kestra instance and create, edit, and run flows.

## Step 4: Use Cloud SQL PostgreSQL as a database backend

This first installation relies on a PostgreSQL database running alongside the Kestra server - on the VM instance (see the PostgreSQL service running thanks to the docker-compose).

For a simple PoC, you can keep that Postgres database running in Docker.

However, for a production-grade installation, we recommend a managed database service such as Cloud SQL.

**Create a Cloud SQL database**

1. Go to the [Cloud SQL console](https://console.cloud.google.com/sql/instances).
2. Click on `Choose PostgreSQL` (Kestra also supports MySQL, but PostgreSQL is recommended).
3. Put an appropriate Instance ID and password for the admin user `postgres`.
4. Select the latest PostgreSQL version from the dropdown.
5. Choose `Enterprise Plus` or `Enterprise` edition based on your requirements.
6. Choose an appropriate preset among `Production`, `Development` or `Sandbox` as per your requirement.
7. Choose the appropriate region and zonal availability.
8. Hit create and wait for completion.
```

![db_choices](/docs/administrator-guide/deployment/gcp-vm/db_choices.png)

![db_setup](/docs/administrator-guide/deployment/gcp-vm/db_setup.png)

> **Enable VM connection to database**
> 
> 1. Go to the database overview page, and click on `Connections` from the left side navigation menu.
> 2. Go to `Networking` tab, and click on `Add a Network`.
> 3. In the New Network section, add an appropriate name like `Kestra VM`, and put your VM instance's external IP address in the Network.
> 4. Click on `Done` in the section.
> 5. Click on `Save` on the page.

Reviewer: ```suggestion
**Enable VM connection to database**

1. Go to the database overview page and click on `Connections` from the left-side navigation menu.
2. Go to the `Networking` tab, and click on `Add a Network`.
3. In the New Network section, add an appropriate name like `Kestra VM`, and put your VM instance's external IP address in the Network.
4. Click on `Done` in the section.
5. Click on `Save` on the page.
```

![db_connections](/docs/administrator-guide/deployment/gcp-vm/db_connections.png)

![db_add_a_network](/docs/administrator-guide/deployment/gcp-vm/db_add_a_network.png)

![db_save_network](/docs/administrator-guide/deployment/gcp-vm/db_save_network.png)

> **Create database user**
> 
> 1. Go to the database overview page, and click on `Users` from the left side navigation menu.
> 2. Click on `Add User Account`.
> 3. Put an appropriate username and password, and click on `Add`.

Reviewer: ```suggestion
**Create database user**

1. Go to the database overview page and click on `Users` from the left-side navigation menu.
2. Click on `Add User Account`.
3. Put an appropriate username and password, and click on `Add`.
```

![db_users](/docs/administrator-guide/deployment/gcp-vm/db_users.png)

![db_user_creation](/docs/administrator-guide/deployment/gcp-vm/db_user_creation.png)

**Create Kestra database**

1. Go to the database overview page, and click on `Databases` from the left side navigation menu.
2. Click on `Create Database`.
3. Put an appropriate database name, and click on `Create`.

**Update Kestra configuration**

In the docker-compose configuration, edit the `datasources` property of the Kestra service in the following way:

```yaml

> datasources:
>     postgres:
>     url: jdbc:postgresql://<your-db-external-endpoint>:5432/<db_name>
>     driverClassName: org.postgresql.Driver
>     username: <your-username>
>     password: <your-password>

Reviewer: ```suggestion
datasources:
  postgres:
  url: jdbc:postgresql://<your-db-external-endpoint>:5432/<db_name>
  driverClassName: org.postgresql.Driver
  username: <your-username>
  password: <your-password>
```

Reviewer: nitpick: YAML uses 2 lines as indentation, 4 is too many

```

> Because you now use RDS service, you don't need the postgres service anymore. Remove it from the `docker-compose.yml` file.
> 
> In order for the changes to take effect, restart the docker services with `docker compose restart` or `docker compose up -d`.
> 
> 
> ## Step 5: Use AWS S3 for storage
> 
> By default, internal storage is implemented using the local file system. This section will guide you on how to change the storage backend to Cloud Storage to ensure a more reliable, durable and scalable storage.
> 
> 1. Go to the Cloud Storage console and create a bucket.
> 2. Go to IAM and select `Service Accounts` from the left side navigation menu.
> 3. On the Service Accounts page, click on `Create Service Account` on the top of the page.
> 4. Put appropriate Service account name and Service account description, and grant the service account `Storage Admin` access. Click Done.
> 5. On the Service Accounts page, click on the newly created service account.
> 6. On the newly created service account page, go to `Keys` tab on the top of the page, and click on `Add Key`. From the dropdown, select `Create New Key`.
> 7. Select Key type as `JSON` and click on `Create`. The JSON key file for the service account will get downloaded.
> 8. We will be using the stringified JSON for our configuration. You can use the bash command `% cat <path_to_json_file> | jq '@json'` to generate stringified JSON.
> 9. Edit Kestra storage configuration.

Reviewer: ```suggestion
Because you now use the Cloud SQL service, you don't need the Postgres Docker service anymore. Remove it from the `docker-compose.yml` file.

In order for the changes to take effect, restart the docker services with `docker compose restart` or `docker compose up -d`.

## Step 5: Use AWS S3 for storage

By default, internal storage is implemented using the local file system. This section will guide you on how to change the storage backend to Cloud Storage to ensure more reliable, durable, and scalable storage.

1. Go to the Cloud Storage console and create a bucket.
2. Go to IAM and select `Service Accounts` from the left-side navigation menu.
3. On the Service Accounts page, click on `Create Service Account` at the top of the page.
4. Put the appropriate Service account name and Service account description, and grant the service account `Storage Admin` access. Click Done.
5. On the Service Accounts page, click on the newly created service account.
6. On the newly created service account page, go to the `Keys` tab at the top of the page and click on `Add Key.` From the dropdown, select `Create New Key`.
7. Select the Key type as `JSON` and click on `Create`. The JSON key file for the service account will get downloaded.
8. We will be using the stringified JSON for our configuration. You can use the bash command `% cat <path_to_json_file> | jq '@json'` to generate stringified JSON.
9. Edit Kestra storage configuration.
```

```yaml
kestra:
  storage:
      type: gcs
      gcs:
        bucket: "<your-cloud-storage-bucket-name>"
        project: "<your-gcp-project-name>"
        serviceAccount: "<stringified-json-file-contents>"
```

In order for the changes to take effect, restart the docker services with `docker compose restart` or `docker compose up -d`.

## Next steps

> This guide walked you through installing Kestra on an GCP VM instance with Cloud SQL database and Cloud Storage as storage backend.

Reviewer: ```suggestion
This guide walked you through installing Kestra on a GCP VM instance with Cloud SQL database and Cloud Storage as storage backend.
```
